tests/Propulsion.MessageDb.Integration/Tests.fs

```diff
@@ -0,0 +1,81 @@
+        for _ in 1..20 do
+            let cmd = NpgsqlBatchCommand()
+            cmd.CommandText <- "select 1 from write_message(@Id::text, @StreamName, @EventType, @Data, 'null', null)"
+            cmd.Parameters.AddWithValue("Id", NpgsqlDbType.Uuid, Guid.NewGuid()) |> ignore
+            cmd.Parameters.AddWithValue("StreamName", NpgsqlDbType.Text, streamName) |> ignore
+            cmd.Parameters.AddWithValue("EventType", NpgsqlDbType.Text, "Hello") |> ignore
+            cmd.Parameters.AddWithValue("Data", NpgsqlDbType.Jsonb, """{"name": "world"}""") |> ignore
+
+            batch.BatchCommands.Add(cmd)
+    do! batch.ExecuteNonQueryAsync() :> Task }
+
+module Array =
+
+    let isAscending arr = Array.sort arr = arr
+
+[<Fact>]
+let ``It processes events for a category`` () = async {
+    let log = Serilog.Log.Logger
+    let consumerGroup = $"{Guid.NewGuid():N}"
+    let category = $"{Guid.NewGuid():N}"
+    do! writeMessagesToCategory category |> Async.AwaitTaskCorrect
+    let reader = MessageDbCategoryReader("Host=localhost; Database=message_store; Port=5433; Username=message_store; Password=;")
+    let checkpoints = ReaderCheckpoint.NpgsqlCheckpointStore("Host=localhost; Database=message_store; Port=5433; Username=postgres; Password=postgres", "public", $"TestGroup{consumerGroup}", TimeSpan.FromSeconds 10)
+    do! checkpoints.CreateSchemaIfNotExists()
+    let stats = { new Propulsion.Streams.Stats<_>(log, TimeSpan.FromMinutes 1, TimeSpan.FromMinutes 1)
+                      with member _.HandleExn(log, x) = ()
+                           member _.HandleOk x = () }
+    let stop = ref (fun () -> ())
+    let handled = HashSet<_>()
+    let handle struct(stream, evts: StreamSpan<_>) = async {
+        lock handled (fun _ -> for evt in evts do handled.Add((stream, evt.Index)) |> ignore)
+        test <@ Array.chooseV Simple.codec.TryDecode evts |> Array.forall ((=) (Simple.Hello {name = "world"})) @>
+        if handled.Count >= 2000 then
+            stop.contents()
+        return struct (Propulsion.Streams.SpanResult.AllProcessed, ()) }
+    use sink = Propulsion.Streams.Default.Config.Start(log, 2, 2, handle, stats, TimeSpan.FromMinutes 1)
+    let source = MessageDbSource(log, TimeSpan.FromMinutes 1, reader, 1000, TimeSpan.FromMilliseconds 100, checkpoints, sink)
+    use src = source.Start(source.Pump(fun _ -> async { return [| TrancheId.parse category |] }))
+    // who says you can't do backwards referencing in F#
+    stop.contents <- src.Stop
+
+    Task.Delay(TimeSpan.FromSeconds 20).ContinueWith(fun _ -> src.Stop()) |> ignore
+
+    do! src.AwaitShutdown()
+    // 2000 total events
+    test <@ handled.Count = 2000 @>
+    // 20 in each stream
+    test <@ handled |> Array.ofSeq |> Array.groupBy fst |> Array.map (snd >> Array.length) |> Array.forall ((=) 20) @>
+    // they were handled in order within streams
+    test <@ handled |> Array.ofSeq |> Array.groupBy fst |> Array.map snd |> Array.forall Array.isAscending @>
```

Comment:
can probably assert each stream's observed `Index` values were `[|1..20|]` ?

Reply:
Yep, that's cleaner!